docs/jep/11859-type-annotations.md

```diff
@@ -0,0 +1,192 @@
+
+![VSCode Intellisense Screenshot](../_static/vscode-completion.png)
+
+Listed in the autocomplete field are all methods and attributes declared by the abstract `ndarray` class.
+We'll discuss further below why it was necessary to create this abstract class rather than annotating with `DeviceArray` directly.
+
+### Level 3: Annotations for static type-checking
+
+These days, static type-checking often is the first thing people think of when considering the purpose of type annotations in Python code.
+While Python does not do any runtime checking of types, several mature static type checking tools exist that can do this as part of a CI test suite.
+The most important ones for JAX are the following:
+
+- [python/mypy](https://github.com/python/mypy) is more or less the standard in the open Python world. JAX currently runs mypy on a subset of source files within the Github Actions CI checks.
+- [google/pytype](https://github.com/google/pytype) is Google's static type checker, and projects which depend on JAX within Google frequently use this.
+- [microsoft/pyright](https://github.com/microsoft/pyright) is important as the static type checker used within VSCode for the Pylance completions mentioned previously.
+
+Full static type checking is the strictest of all the type annotation applications, because it will surface an error any time your type annotations are not precisely correct.
+On the one hand, this is nice because your static type analysis may catch faulty type annotations (for example, a case where a `DeviceArray` method is missing from the `jnp.ndarray` abstract class).
+
+On the other hand, this strictness can make the type checking process very brittle in packages that often rely on duck-typing rather than strict type-safe APIs.
+You'll currently find code comments like `#type: ignore` (for mypy) or `#pytype: disable` (for pytype) peppered throughout the JAX codebase in several hundred places.
+These typically represent cases where typing problems have arisen; they may be inaccuracies in JAX type annotations, or inaccuracies in the static type checker's ability to correctly follow the control flow in the code.
+On occasion, they are due to real & subtle bugs in the behavior of pytype or mypy.
+In rare cases, they may be due to the fact that JAX uses Python patterns that are difficult or even impossible to express in terms of Python's static type annotation syntax.
+
+## Type annotation challenges for JAX
+
+JAX currently has type annotations that are a mixture of different styles, and aimed at all three levels of type annotation discussed above.
+Partly, this comes from the fact that JAX's source code poses a number of unique challenges for Python's type annotation system. We'll outline them here.
+
+### Challenge 1: pytype, mypy and developer friction
```

Comment:
I think the friction is actually a result of Python not having a precise description of what the type system is and how it is supposed to work. So, mypy, pytype, pyright, pyre, IntelliJ etc all implement a slightly different type system.

@kmillikin did a talk about this on the recent Typing Summit: https://youtu.be/BNTkWQfqP_c?t=8466

Reply:
Thanks for the link! That's a good point – but I think JAX faces unique friction here because we use two different type-checkers in our CI, and different populations of JAX developers have access to only one or the other while writing their code.